cpp/src/strings/repeat_strings.cu

```diff
@@ -172,14 +167,204 @@ std::unique_ptr<column> repeat_strings(strings_column_view const& input,
+
+    auto [offsets_column, chars_column] =
+      make_strings_children(fn, strings_count, strings_count, output_strings_offsets, stream, mr);
+
+    // If only one input column has nulls, we just copy its null mask and null count.
+    // If both input columns have nulls, we generate new bitmask by AND their bitmasks.
+    auto [null_mask, null_count] = [&] {
+      if (strings_has_nulls ^ rtimes_has_nulls) {
+        auto const& col = strings_has_nulls ? input.parent() : repeat_times;
+        return std::make_pair(cudf::detail::copy_bitmask(col, stream, mr), col.null_count());
+      } else if (strings_has_nulls && rtimes_has_nulls) {
+        return std::make_pair(
+          cudf::detail::bitmask_and(table_view{{input.parent(), repeat_times}}, stream, mr),
+          UNKNOWN_NULL_COUNT);
+      }
+      return std::make_pair(rmm::device_buffer{0, stream, mr}, 0);
+    }();
+
+    return make_strings_column(strings_count,
+                               std::move(offsets_column),
+                               std::move(chars_column),
+                               null_count,
+                               std::move(null_mask),
+                               stream,
+                               mr);
+  }
+
+  template <class T, std::enable_if_t<!cudf::is_index_type<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(size_type,
+                                     strings_column_view const&,
+                                     column_view const&,
+                                     std::optional<column_view>,
+                                     rmm::cuda_stream_view,
+                                     rmm::mr::device_memory_resource*) const
+  {
+    CUDF_FAIL("repeat_strings expects an integer type for the `repeat_times` input column.");
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> repeat_strings(strings_column_view const& input,
+                                       column_view const& repeat_times,
+                                       std::optional<column_view> output_strings_offsets,
+                                       rmm::cuda_stream_view stream,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(input.size() == repeat_times.size(), "The input columns must have the same size.");
+  CUDF_EXPECTS(!output_strings_offsets.has_value() ||
+                 input.size() + 1 == output_strings_offsets.value().size(),
+               "Invalid output strings offsets.");
+
+  auto const strings_count = input.size();
+  if (strings_count == 0) { return make_empty_column(data_type{type_id::STRING}); }
+
+  return type_dispatcher(repeat_times.type(),
+                         dispatch_repeat_strings_separately_fn{},
+                         strings_count,
+                         input,
+                         repeat_times,
+                         output_strings_offsets,
+                         stream,
+                         mr);
```

Comment:
Since `repeat_times` is expected to be integers (i.e. index-type) you can use the [`indexalator`](https://github.com/rapidsai/cudf/blob/branch-21.08/cpp/include/cudf/detail/indexalator.cuh#L217-L242) here instead of the type-dispatcher. The `indexalator` normalizes any integer column type to `size_type` -- you do not need to template the functor. Saves alot of code generation and compile time.